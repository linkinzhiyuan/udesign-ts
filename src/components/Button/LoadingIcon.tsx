import React from "react";
import CSSMotion from 'rc-motion'
import LoadingOutlined from '@ant-design/icons/LoadingOutlined'

export interface LoadingIconProps {
  prefixCls?: string;
  existIcon?: boolean;
  loading?: boolean | object
}

const getCollapsedWidth = () => ({ width: 0, opacity: 0, transform: 'scale(0)'});

const getRealWidth = (node: HTMLElement) => {
  const width = node.scrollWidth;  
  return { width, opacity: 1, transform: 'scale(1)'};
};

const LoadingIcon: React.FC<LoadingIconProps> = ({ prefixCls, loading, existIcon}) => {
  const visible = !!loading;

  if(existIcon) {
    return (<span className={`${prefixCls}-loading-icon`}>
      <LoadingOutlined onPointerEnterCapture={undefined} onPointerLeaveCapture={undefined} />
    </span>)
  }

  return (
    <CSSMotion 
      visible={visible}
      motionName={`${prefixCls}-loading-icon-motion`}
      removeOnLeave
      onAppearStart={getCollapsedWidth}
      onAppearActive={getRealWidth}
      onLeaveStart={getRealWidth}
      onLeaveActive={getCollapsedWidth}
      onEnterActive={getRealWidth}
      onEnterStart={getCollapsedWidth}
    >
      {({className, style}: {className?: string, style?: React.CSSProperties}, ref: any) => (
        <span className={`${prefixCls}-loading-icon`} style={style} ref={ref}>
          <LoadingOutlined  onPointerEnterCapture={undefined} onPointerLeaveCapture={undefined} className={className}/>
        </span>
      )}
    </CSSMotion>
  )
}

export default LoadingIcon
