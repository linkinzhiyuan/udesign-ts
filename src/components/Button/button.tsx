import React from 'react';
import cls from 'classnames';
import omit from 'rc-util/lib/omit';
import './style/index.less';

const ButtonTypes = ['primary', 'default', 'ghost','dashed', 'text', 'link'] as const;
type ButtonType = typeof ButtonTypes[number];

const shapeMap = ['circle', 'round'] as const;
type ShapeType = typeof shapeMap[number];

const sizeMap = ['small', 'default', 'large'] as const;
type SizeType = typeof sizeMap[number];

const HTMLTypes = ['submit', 'reset', 'button'] as const;
type HTMLType = typeof HTMLTypes[number];

export interface BaseButtonProps {
  type?: ButtonType,
  icon?: React.ReactNode,
  shape?: ShapeType,
  size?: SizeType,
  disabled?: boolean,
  loading?: boolean | { delay?: number },
  prefixCls?: string,
  className?: string,
  ghost?: boolean,
  danger?: boolean,
  block?: boolean,
  children?: React.ReactNode,
}

// 链接
export type AnchorButtonProps = {
  href: string;
  target?: string;
  onClick?: React.MouseEventHandler<HTMLElement>;
} & BaseButtonProps & Omit<React.AnchorHTMLAttributes<HTMLAnchorElement>, 'type' | 'onClick'>

// 原生按钮
export type NativeButtonProps = {
  htmlType?: HTMLType;
  onClick?: React.MouseEventHandler<HTMLElement>;
} & BaseButtonProps & Omit<React.ButtonHTMLAttributes<HTMLElement>, 'type' | 'onClick'>;


export type ButtonProps = Partial<NativeButtonProps & AnchorButtonProps>

type Loading = number | boolean

const InternalButton: React.ForwardRefRenderFunction<unknown,ButtonProps> = (props, ref) => {
  const { loading = false,type = 'default', disabled: customDisabled, size: customizeSize, className,prefixCls, children, ...rest } = props;
  const buttonRef = ref as any || React.createRef<HTMLElement>();

  // ========================= Disabled ==========================
  const disabled = customDisabled; // TODO: 全局Disabled
  const mergeDisabled = customDisabled ?? disabled;


  const [innerLoading, setLoading] = React.useState<Loading>(!!loading);

  const linkButtonRestProps = omit(rest as AnchorButtonProps & { navigate: any}, ['navigate']);
  // 点击事件
  const handleClick = (e: React.MouseEvent<HTMLButtonElement | HTMLAnchorElement, MouseEvent>) => {
    const { onClick } = props;

    if(innerLoading || mergeDisabled) {
      e.preventDefault();
      return;
    }
    (onClick as React.MouseEventHandler<HTMLButtonElement | HTMLAnchorElement>)?.(e);
  }

  // ========================= update Loading ==========================
  const loadingOrDelay = typeof loading === 'boolean' ? loading : loading?.delay || true

  React.useEffect(() => {
    let delayTimer: number | null = null;

    if(typeof loadingOrDelay === 'number') {
      delayTimer = window.setTimeout(() => {
        delayTimer = null;
        setLoading(loadingOrDelay);
      }, loadingOrDelay);
    } else {
      setLoading(loadingOrDelay)
    }

    return () => {
      if(delayTimer) {
        window.clearTimeout(delayTimer);
        delayTimer = null;
      }
    }
  }, [loadingOrDelay])


  // className
  const classes = cls(
    prefixCls,
    'uDesign-btn',
    `uDesign-btn-${type}`,
    `uDesign-btn-${customizeSize}`,
    mergeDisabled && 'uDesign-btn--disabled',
    className,
  )

  const iconNode = <></>

  // 存在链接地址即为 a标签
  if (linkButtonRestProps.href !== undefined) {
    return (
      <a {...linkButtonRestProps} className={classes} onClick={handleClick} ref={buttonRef}>
        {iconNode}
        {children}
      </a>
    )
  }

  // 正常的按钮
  return (
    <button {...(rest as NativeButtonProps) } className={classes} onClick={handleClick} type='button' disabled={disabled} ref={buttonRef}>
      {iconNode}
      {children}
    </button>
  )
};

type CompoundedComponent = React.ForwardRefExoticComponent<ButtonProps & React.RefAttributes<unknown>> & {
  // Group: typeof ButtonGroup;
  __ANT_BUTTON: boolean;
}

const Button = React.forwardRef<unknown, ButtonProps>(InternalButton) as CompoundedComponent
export { Button }
