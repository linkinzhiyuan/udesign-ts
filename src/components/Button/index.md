---
category: Components
title: Button
subtitle: 按钮
description: 按钮用于开始一个即时操作。
cover: https://mdn.alipayobjects.com/huamei_7uahnr/afts/img/A*7va7RKs3YzIAAAAAAAAAAAAADrJ8AQ/original
coverDark: https://mdn.alipayobjects.com/huamei_7uahnr/afts/img/A*3T4cRqxH9-8AAAAAAAAAAAAADrJ8AQ/original
demo:
  cols: 2
group:
  title: 组件
  order: 1
---

## 代码演示

<!-- prettier-ignore -->
<code src="./demo/basic.tsx">语法糖</code>
<!-- <code src="./demo/color-variant.tsx" version="5.21.0">颜色与变体</code>
<code src="./demo/debug-color-variant" debug>调试颜色与变体</code>
<code src="./demo/icon.tsx">按钮图标</code>
<code src="./demo/icon-position.tsx" version="5.17.0">按钮图标位置</code>
<code src="./demo/debug-icon.tsx" debug>调试图标按钮</code>
<code src="./demo/debug-block.tsx" debug>调试按钮block属性</code>
<code src="./demo/size.tsx">按钮尺寸</code>
<code src="./demo/disabled.tsx">不可用状态</code>
<code src="./demo/loading.tsx">加载中状态</code>
<code src="./demo/multiple.tsx">多个按钮组合</code>
<code src="./demo/ghost.tsx">幽灵按钮</code>
<code src="./demo/danger.tsx">危险按钮</code>
<code src="./demo/block.tsx">Block 按钮</code>
<code src="./demo/legacy-group.tsx" debug>废弃的 Block 组</code>
<code src="./demo/chinese-chars-loading.tsx" debug>加载中状态 bug 还原</code>
<code src="./demo/component-token.tsx" debug>组件 Token</code>
<code src="./demo/linear-gradient.tsx">渐变按钮</code>
<code src="./demo/noSpace.tsx" version="5.17.0">移除两个汉字之间的空格</code> -->
