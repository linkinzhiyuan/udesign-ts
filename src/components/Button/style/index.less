@import '../../../style/themes/variable.less';
@import './mixin.less';

.uDesign-btn {
  display: inline-block;
  line-height: 1;
  white-space: nowrap;
  cursor: pointer;
  background: #fff;
  border: 1px solid #d9d9d9;
  border-color: #d9d9d9;
  color: rgba(0,0,0,.65);
  -webkit-appearance: none;
  text-align: center;
  box-sizing: border-box;
  outline: 0;
  margin: 0;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  padding: 0 15px;
  font-size: 14px;
  border-radius: 4px;
  height: 32px;

  &-link {
    .btn-link();
    // border: none;
  }
}

.uDesign-btn-primary {
  color: #fff;
  background-color: #1890ff;
  border-color: #1890ff;
  text-shadow: 0 -1px 0 rgba(0,0,0,.12);
  box-shadow: 0 2px 0 rgba(0,0,0,.045);
}

.uDesign-btn-primary:hover {
  color: #fff;
  background: #40a9ff;
  border-color: #40a9ff;
}

