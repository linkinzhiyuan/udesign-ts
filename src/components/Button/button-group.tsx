import cls from 'classnames';
import React from 'react';
import { getPrefixCls } from './button';
import warning from '../_util/warning';

export interface ButtonGroupProps {
  prefixCls?: string;
  className?: string;
  children?: React.ReactNode;
  style?: React.CSSProperties;
  size?: 'small' | 'middle' | 'large' | undefined;
}

const ButtonGroup: React.FC<ButtonGroupProps> = (props) => {
  const { className, prefixCls: customizePrefixCls, size, children, ...restProps } = props;

  let sizeCls = '';
  switch (size) {
    case 'large':
      sizeCls = 'lg';
      break;
    case 'small':
      sizeCls = 'sm';
      break;
    case 'middle':
    case undefined:
      break;
    default:
      warning(!size, 'Button.Group', 'Invalid prop `size`.')
      
  }

  const prefixCls = getPrefixCls('btn-group', customizePrefixCls);

  const classes = cls(
    prefixCls,
    {
      [`${prefixCls}-${sizeCls}`]: sizeCls,
    },
    className
  )
  return (
    <div {...restProps} className={classes}>
      {children}
    </div>
  );
};

export default ButtonGroup;
