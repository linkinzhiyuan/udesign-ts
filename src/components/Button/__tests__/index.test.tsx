import React, { Component } from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import Button from '../index'; // 假设 Button 组件位于同一目录下
import '@testing-library/jest-dom';


// import { SearchOutlined } from '@ant-design/icons';
import { resetWarned } from 'rc-util/lib/warning';
// import { act } from 'react-dom/test-utils';
import mountTest from '../../../tests/mountTest';
import rtlTest from '../../../tests/rtlTest';
// import { fireEvent, render, sleep } from '../../../tests/utils';
// import ConfigProvider from '../../config-provider';

describe('Button Component', () => {
  it('renders correctly and responds to click events', () => {
    // 创建一个模拟的点击处理函数
    const handleClick = jest.fn();

    // 渲染 Button 组件
    render(<Button onClick={handleClick}>Click Me</Button>);

    // 查找按钮元素
    const buttonElement = screen.getByRole('button', { name: /click me/i });

    // 验证按钮是否在文档中
    expect(buttonElement).toBeInTheDocument();

    // 模拟点击事件
    fireEvent.click(buttonElement);

    // 验证点击处理函数是否被调用
    expect(handleClick).toHaveBeenCalledTimes(1);
  });
});

describe('Button', () => {
  mountTest(Button);
  mountTest(() => <Button size="large" />);
  mountTest(() => <Button size="small" />);
  // mountTest(Button.Group);
  // mountTest(() => <Button.Group size="large" />);
  // mountTest(() => <Button.Group size="small" />);
  // mountTest(() => <Button.Group size="middle" />);

  rtlTest(Button);
  rtlTest(() => <Button size="large" />);
  rtlTest(() => <Button size="small" />);
  // rtlTest(Button.Group);
  // rtlTest(() => <Button.Group size="large" />);
  // rtlTest(() => <Button.Group size="small" />);
  // rtlTest(() => <Button.Group size="middle" />);

  // it('renders correctly', () => {
  //   const { container } = render(<Button>Follow</Button>);
  //   expect(container.firstChild).toMatchSnapshot();
  // });

  it('mount correctly', () => {
    expect(() => render(<Button>Follow</Button>)).not.toThrow();
  });

  // it('warns if size is wrong', () => {
  //   resetWarned();
  //   const mockWarn = jest.spyOn(console, 'error').mockImplementation(() => {});
  //   const size = 'who am I';
  //   // @ts-expect-error: Type '"who am I"' is not assignable to type 'SizeType'.ts(2322)
  //   render(<Button.Group size={size} />);
  //   expect(mockWarn).toHaveBeenCalledWith('Warning: [antd: Button.Group] Invalid prop `size`.');

  //   mockWarn.mockRestore();
  // });

  // it('renders Chinese characters correctly', () => {
  //   expect(render(<Button>按钮</Button>).container.firstChild).toMatchSnapshot();
  //   // should not insert space when there is icon
  //   expect(
  //     render(<Button icon={<SearchOutlined />}>按钮</Button>).container.firstChild,
  //   ).toMatchSnapshot();
  //   // should not insert space when there is icon
  //   expect(
  //     render(
  //       <Button>
  //         <SearchOutlined />
  //         按钮
  //       </Button>,
  //     ).container.firstChild,
  //   ).toMatchSnapshot();
  //   // should not insert space when there is icon
  //   expect(
  //     render(<Button icon={<SearchOutlined />}>按钮</Button>).container.firstChild,
  //   ).toMatchSnapshot();
  //   // should not insert space when there is icon while loading
  //   expect(
  //     render(
  //       <Button icon={<SearchOutlined />} loading>
  //         按钮
  //       </Button>,
  //     ).container.firstChild,
  //   ).toMatchSnapshot();
  //   // should insert space while loading
  //   expect(render(<Button loading>按钮</Button>).container.firstChild).toMatchSnapshot();

  //   // should insert space while only one nested element
  //   expect(
  //     render(
  //       <Button>
  //         <span>按钮</span>
  //       </Button>,
  //     ).container.firstChild,
  //   ).toMatchSnapshot();
  // });

  // it('renders Chinese characters correctly in HOC', () => {
  //   const Text = ({ children }: { children: React.ReactNode }) => <span>{children}</span>;
  //   const { container, rerender } = render(
  //     <Button>
  //       <Text>按钮</Text>
  //     </Button>,
  //   );
  //   expect(container.querySelector('.ant-btn')).toHaveClass('ant-btn-two-chinese-chars');

  //   rerender(
  //     <Button>
  //       <Text>大按钮</Text>
  //     </Button>,
  //   );
  //   expect(container.querySelector('.ant-btn')).not.toHaveClass('ant-btn-two-chinese-chars');

  //   rerender(
  //     <Button>
  //       <Text>按钮</Text>
  //     </Button>,
  //   );
  //   expect(container.querySelector('.ant-btn')).toHaveClass('ant-btn-two-chinese-chars');
  // });

  // // https://github.com/ant-design/ant-design/issues/18118
  // it('should not insert space to link or text button', () => {
  //   const wrapper1 = render(<Button type="link">按钮</Button>);
  //   expect(wrapper1.getByRole('button')).toHaveTextContent('按钮');
  //   wrapper1.unmount();
  //   const wrapper2 = render(<Button type="text">按钮</Button>);
  //   expect(wrapper2.getByRole('button')).toHaveTextContent('按钮');
  // });

  // it('should render empty button without errors', () => {
  //   const wrapper = render(
  //     <Button>
  //       {null}
  //       {undefined}
  //     </Button>,
  //   );
  //   expect(wrapper.container.firstChild).toMatchSnapshot();
  // });

  it('have static property for type detecting', () => {
    expect(Button.__ANT_BUTTON).toBe(true);
  });

  // it('should warning when pass a string as icon props', () => {
  //   resetWarned();
  //   const warnSpy = jest.spyOn(console, 'error').mockImplementation(() => {});

  //   render(<Button type="primary" icon="ab" />);
  //   expect(warnSpy).not.toHaveBeenCalled();

  //   render(<Button type="primary" icon="search" />);
  //   expect(warnSpy).toHaveBeenCalledWith(
  //     `Warning: [antd: Button] \`icon\` is using ReactNode instead of string naming in v4. Please check \`search\` at https://ant.design/components/icon`,
  //   );

  //   warnSpy.mockRestore();
  // });

  // it('should warning when pass type=link and ghost=true', () => {
  //   resetWarned();
  //   const warnSpy = jest.spyOn(console, 'error').mockImplementation(() => {});
  //   render(<Button type="link" ghost />);
  //   expect(warnSpy).toHaveBeenCalledWith(
  //     "Warning: [antd: Button] `link` or `text` button can't be a `ghost` button.",
  //   );
  //   warnSpy.mockRestore();
  // });

  // it('should warning when pass type=text and ghost=true', () => {
  //   resetWarned();
  //   const warnSpy = jest.spyOn(console, 'error').mockImplementation(() => {});
  //   render(<Button type="text" ghost />);
  //   expect(warnSpy).toHaveBeenCalledWith(
  //     "Warning: [antd: Button] `link` or `text` button can't be a `ghost` button.",
  //   );
  //   warnSpy.mockRestore();
  // });


});
