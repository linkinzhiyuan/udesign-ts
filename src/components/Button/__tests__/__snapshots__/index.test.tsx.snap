// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button rtl render component should be rendered correctly in RTL direction 1`] = `
<button
  class="uDesign-btn uDesign-btn-default uDesign-btn-undefined"
  type="button"
/>
`;

exports[`Button rtl render component should be rendered correctly in RTL direction 2`] = `
<button
  class="uDesign-btn uDesign-btn-default uDesign-btn-large"
  type="button"
/>
`;

exports[`Button rtl render component should be rendered correctly in RTL direction 3`] = `
<button
  class="uDesign-btn uDesign-btn-default uDesign-btn-small"
  type="button"
/>
`;
